### ISSUE TITLE
New ADT structure including checking

### ISSUE THREAD

**Issue by pjljvandelaar** [TorXakis/TorXakis#565]
- [ ] Data strucutre for ADT
- [ ] Checking / validation of data structure (when and what)
- [ ] test : generator for ADT (in Quickcheck using sized)
- [ ] test: tests for ADT
- [ ] documentation: haddock documentation
     - [ ] make interface of ValExpr w.r.t. ADT clear
- [ ] follow changes in higher levels of TorXakis

estimate time: 3 weeks

**Comment by pjljvandelaar** [comment]
I propose to use smart constructors: see https://wiki.haskell.org/Smart_constructors for more info

**Comment by pjljvandelaar** [comment]
See #503 for an additional requirement

**Comment by pjljvandelaar** [comment]
To make strings as short as possible use integers in another base: see 
https://stackoverflow.com/questions/1959715/how-to-print-integer-literals-in-binary-or-hex-in-haskell

Could one also have a base large than 16?

**Comment by pjljvandelaar** [comment]
About the name SORT
https://dictionary.cambridge.org/grammar/british-grammar/common-nouns/sort-type-and-kind
http://difference-between.com/english-language/kind-sort-and-type

define:sort yields
1. a category of things or people with a common feature; a type.
2. COMPUTING: the arrangement of data in a prescribed sequence.

define:type yields
1. a category of people or things having common characteristics.
2. a person or thing exemplifying the ideal or defining characteristics of something.
3. characters or letters that are printed or shown on a screen.
4. a design on either side of a medal or coin.
5. LINGUISTICS: an abstract category or class of linguistic item or unit, as distinct from actual occurrences in speech or writing.

**Comment by pjljvandelaar** [comment]
@tretmans could you point to a document/websites that makes a difference between sort and type: with and without operators and functions?

**Comment by pjljvandelaar** [comment]
@keremispirli @capitanbatata Could we use https://wiki.haskell.org/Mutually_recursive_modules to break the cycle of  Sort and ADTDEF?

**Comment by keremispirli** [comment]
@pjljvandelaar Looks like we can break the cycle by changing `SortADT (Ref ADTDef)` to `SortADT (Ref t)`, sacrificing type-safety for SortADT.

Other option is writing a `.hs-boot` file, about which I have no idea.

**Comment by pjljvandelaar** [comment]
I thought we tried `SortADT (Ref t)` then
`Sort` would also get a `t` and ADTDef that uses sort would also get a `t`.....

So it would become `Ref ADTDEF Sort` (maybe with brackets) which I don't consider readable....

**Comment by keremispirli** [comment]
Experiment results:

Using `Ref Int` introduces a lot of complexity when creating and accessing data structures. Yet it does not provide significant performance gains against using HashMap with entity names (Text).

Test results: Access 1m values with Map lookups

- Map     -> 1.8s
- IntMap -> 0.5s
- HashMap -> 0.7s

Using names (as Text) as references also removes the need of looking up data structure just for its name when name is the only thing that is needed.

**Comment by pjljvandelaar** [comment]
We have two different ways to specify Sort:
* The user uses `Name`: it is unchecked whether this name is valid.
* Internally we use `Sort (Ref ADTDef)`: the reference is checked and is either a predefined sort, or points to a declared ADT.

This mean that we have a number of data structure where both ways to specify sort are possible
* `FieldDef v`
* `FieldDefs v`
* `ConstructorDef v`
* `ConstructorDefs v`
* `ADTDef v`

When we make `ADTDefs` the reference are checked and we no longer need the distinction between user specified and still unchecked values and internal representation with checked values!

The transformation from `X Name` to `X (Sort (Ref ADTDef))` can be realized using [https://wiki.haskell.org/Scrap_your_boilerplate]([scrap your boilerplate](url))
See also http://expressiveprogramming.com/presentations/syb_talk.html

**Comment by pjljvandelaar** [comment]
So the code for `FieldDef v` and `FieldDefs v` becomes
```
-- | Data structure for a field definition.
data FieldDef v = FieldDef { fieldName :: Text -- ^ Name of the field
                           , sort      :: v    -- ^ Sort of the field
                           }
    deriving (Eq,Ord,Read,Show, Generic, NFData, Data)

-- | Data structure for a collection of 'FieldDef's.
data FieldDefs v = FieldDefs  { -- | Transform 'FieldDefs' to a list of tuples of 'Ref' 'FieldDef' and 'FieldDef'.
                               fDefsToList :: [(Ref (FieldDef v), FieldDef v)]
                                -- | Number of field definitions in a 'FieldDefs'.
                              , nrOfFieldDefs :: Int
                              }
    deriving (Eq, Ord, Read, Show, Generic, NFData, Data)
```

**Comment by keremispirli** [comment]
@pjljvandelaar I don't know whether we have a convention about it yet; it looks like right now we have both.

Comparing current definitions above with what you wrote, are we replacing `Sort` altogether with type parameter v ?

**Comment by pjljvandelaar** [comment]
How to handle `v` in `Ref (ADTDef v)`?
See 
https://wiki.haskell.org/Empty_type and
https://stackoverflow.com/questions/16892570/what-is-in-haskell-exactly

or just use `Ref ADTDefs`?

**Comment by pjljvandelaar** [comment]
We are creating variants: the non-checked (`v` equals `Name`) and the checked (`v` equals `Ref (ADTDef v)`)

**Comment by pjljvandelaar** [comment]
Indeed, `FieldDef Name` contains unchecked sort references (user input), ie, `Name`, while
`FieldDef Sort` contains checked sort references, ie., `Sort` that internally contains `Ref (ADTDef v)`.
Since by defining `ADTDefs` everything is checked, one doesn't need the variant anymore at that level.

**Comment by pjljvandelaar** [comment]
If we would write `Ref (ADTDef Sort)`, I think we have the solution as mentioned on https://wiki.haskell.org/Mutually_recursive_modules

**Comment by pjljvandelaar** [comment]
@keremispirli  about the convention: I think we should use `-- | ` when the constructor is NOT public and `-- ^ ` when it is!

**Comment by capitanbatata** [comment]
@pjljvandelaar @keremispirli I've updated the issue by adding some checkboxes with the tasks I understand need to be finished in order to complete this task. @pjljvandelaar Please take a look at these. Maybe we can merge the checkboxes into one.

**Comment by capitanbatata** [comment]
> @keremispirli about the convention: I think we should use for the fields/accessors -- | when the constructor is NOT public and -- ^ when it is!

Why not just using `--|` everywhere? In that way we have more characters to fit in one line.

**Comment by pjljvandelaar** [comment]
For me '--|' is as long as '--^' ... so I don 't understand your remark....

**Comment by capitanbatata** [comment]
> For me '--|' is as long as '--^' ... so I don 't understand your remark....

You use `-- ^` to refer to the element that precede the comment, and `-- |` to refer to elements that follow. Example:

```haskell
data D = D 
  { f0 :: A -- ^ First field
    -- | Second field
  , f1 :: B
  }
```

**Comment by pjljvandelaar** [comment]
See http://haskell-haddock.readthedocs.io/en/latest/markup.html ...

**Comment by capitanbatata** [comment]
> See http://haskell-haddock.readthedocs.io/en/latest/markup.html ...

Ok, you meant the top level declaration. Why not stick to `-- |`? Less cognitive load (than having to remember conventions about whether it is a private or public function), and nothing needs to be changed if a function is changed to public or vice-versa.

**Comment by pjljvandelaar** [comment]
I agree just stick to `-- |` (had to read the docs to get the real meaning of `-- ^`).

**Comment by capitanbatata** [comment]
I built and installed `argon` on Linux, and it seems to work. See for instance the `lpe` dode:

![screenshot from 2018-02-13 11-08-39](https://user-images.githubusercontent.com/175315/36144433-4a7b1be4-10ae-11e8-8fe9-6a601989e32a.png)

Sadly I think it does not work on Windows, but I can run the analysis and report it here.

**Comment by capitanbatata** [comment]
I'm adding an extra task to move the `.cabal` file to use `hpack` since that will simplify our maintenance tasks.

**Comment by pjljvandelaar** [comment]
Why not make a separate issue: move to hpack?
We can then discuss the pros and cons.

Furthermore it keeps this issue more focussed.

**Comment by capitanbatata** [comment]
I already migrated to `hpack` 🙃 

New projects in stack come with `hpack` by default. See if you like the new approach, we can revert later. But `hpack` is seen as a superior alternative to `cabal` (hpack generates cabal files automatically).

**Comment by capitanbatata** [comment]
>  For example, can we still use 'weeder'?
Yes.

**Comment by capitanbatata** [comment]
@pjljvandelaar and @keremispirli I've finished an alternative version of what you guys did so far. In the `ADTExp.hs` file you can find the untidy version. In the `Alt` folder you can find the more organized version. 

Since @pjljvandelaar starts early on Monday, he can take a look at what I've done already. On Monday after 9:30 the three of us can discuss my questions and suggested changes, and merge what we think appropriate into the work you've done. Hopefully we can all agree, and move forward to try to merge the changes into TorXakis, and finish this story.

To play with the examples, fire up the `ghci` REPL selecting the `sort-test` as target:

```sh
stack ghci --ghc-options -Wno-all sort:test:sort-test
```

Then take a look at:

```haskell
Alt.Examples.mADTDefs 
-- or
Alt.Examples.mCircularADTDefs 
```

The examples are in the `test` folder (`sys/sort/test/Alt/Examples.hs`). For instance:

```haskell
tPerson = "Person" .::= ["Person" .= ["name" .: "String", "age" .: "Int"]]
tOperation = "Operation" .::= [ "Plus"  .= ["p0" .: "Int", "p1" .: "Int"]
                              , "Minus" .= ["m0" .: "Int", "m1" .: "Int"]
                              ]
tStudent = "Student" .::= [ "Student" .= ["who" .: "Person", "grade" .: "Int"]]
mADTDefs = mkADTs' [tPerson, tOperation, tStudent] 
```

**Comment by capitanbatata** [comment]
ADT generators are ready. We should be able to quick-check good weather scenarios. @pjljvandelaar please check if this is what you had in mind.

**Comment by pjljvandelaar** [comment]
@capitanbatata I think you forgot to check in ;-)
I get build errors.
``

**Comment by pjljvandelaar** [comment]
`f:fs <- elements famousPeople` assumes that although `O` is a real last name, `O` will not become famous ;-)

**Comment by pjljvandelaar** [comment]
```
-- The first list of constructors must contain at least one element, and
    -- may not use the type of the containing ADT.
    cs0 <- listOf1 $ arbitraryConstructorDef Nothing
```
I would have create just one element `c0` instead of a list, but both are fine.

**Comment by pjljvandelaar** [comment]
Maybe it is for later, but I am missing some termination logic needed for making `ValExpr` that use constructors and accessors (and which have to terminate despite recursive data types).

**Comment by pjljvandelaar** [comment]
Final remark:
In arbitraryConstructorDefs, I read
```
    -- Clear the field names.
    lift $ modify clearFields
```
I assume this explicitly uses the way the generator works: from the `ConstructorDefs` a single `ADT` is made (next). And never is a single `ADT` made from multiple `ConstructorDefs`.

**Comment by pjljvandelaar** [comment]
@keremispirli We have to discuss `Referencable` once more.
We have `ref by name` `ref by signature` and `ref by unique id`.
For `ref by unique id` we need a Monad (to remember the look up table).

We as developers have a choice: how do we refer?
For example, a reference to a function can be `ref by signature` and `ref by unique id`.
So the `convertsTo`  for
```
instance (Referencable a', ConvertsTo a a') => ConvertsTo [a] (HashMap (Ref a') a') where
    convertTo as =
        Map.fromList
        $ map ((\a' -> (mkRef a', a')) . convertTo) as
```
might be too prescribing: it should be an explicit decision by us (the developers) based on the results of some benchmarks.

**Comment by capitanbatata** [comment]
>  I get build errors.

That's because I did not remove the warnings yet. Try with:

> stack build --fast --file-watch --ghc-options -Wno-all sort-gen             

And to play with the generators load the REPL with:

> stack ghci --ghc-options -Wno-all sort-gen:lib

**Comment by pjljvandelaar** [comment]
@capitanbatata what is generated looks good!

Since I only see [a-zA-Z] as characters, we might not be testing it as thoroughly as we want since I think more characters are allowed to make valid names. Or should we limited what is valid in the smart constructor of `Name`?

**Comment by capitanbatata** [comment]
> Since I only see [a-zA-Z] as characters, we might not be testing it as thoroughly as we want since I think more characters are allowed to make valid names. Or should we limited what is valid in the smart constructor of Name?

I don't know if we want to generate strings with any  printable characters at this point. In the sort package we only use the name for string comparison, so I don't know if using random strings will help us in testing. In the future we might add a switch to toggle between readable names and more random stuff.

**Comment by pjljvandelaar** [comment]
let's discuss on Monday the composability of generators.
How would you generate a function definition that has arguments including adt as sort?

Could we also have made a field generator that would get a set of available data dataypes and a set of forbidden field names, and similar a constructor generator and adt generator?

**Comment by pjljvandelaar** [comment]
I realized that the generators allow recursive data types but only to themselves: like List
```
  List ::= Nil
           | Cstr (head :: Int, tail :: List)
```
I think a group of recursive data types is not generated,
For example,
```
A ::= C | D (b :: B)
B ::= E (a :: A) 
```
is imho not covered.
@capitanbatata Do you agree?

**Comment by capitanbatata** [comment]
@pjljvandelaar agree.  I'll make a new checkbox out of it. @keremispirli please bear in mind that we don't need to complete all the TODO's in the issue description to close https://github.com/TorXakis/TorXakis/pull/585, let's aim to finish the absolute minimum, and promote the unfinished tasks as TODO's.

**Comment by pjljvandelaar** [comment]
When we want to merge this branch with the main development, we have a lot of issues
```
PS C:\TorXakis.git> git merge develop
Auto-merging test/sqatt/stack_linux.yaml
Removing test/sqatt/src/Integration/Any.hs
Removing test/sqatt/data/integration-test/Any/anyTest.txscmd
Removing test/sqatt/data/integration-test/Any/anyTest.txs
Auto-merging test/environment/environment.cabal
CONFLICT (content): Merge conflict in test/environment/environment.cabal
Auto-merging sys/value/value.cabal
Auto-merging sys/value/src/Eval.hs
CONFLICT (content): Merge conflict in sys/value/src/Eval.hs
Auto-merging sys/valexpr/valexpr.cabal
CONFLICT (content): Merge conflict in sys/valexpr/valexpr.cabal
Auto-merging sys/valexpr/test/ProductGen.hs
Auto-merging sys/valexpr/src/ValExprImpls.hs
CONFLICT (content): Merge conflict in sys/valexpr/src/ValExprImpls.hs
Auto-merging sys/valexpr/src/ValExprDefs.hs
CONFLICT (content): Merge conflict in sys/valexpr/src/ValExprDefs.hs
CONFLICT (modify/delete): sys/valexpr/src/SortId.hs deleted in HEAD and modified in develop. Version develop of sys/valexpr/src/SortId.hs left in tree.
Auto-merging sys/valexpr/src/FuncDef.hs
CONFLICT (content): Merge conflict in sys/valexpr/src/FuncDef.hs
CONFLICT (modify/delete): sys/valexpr/src/CstrDef.hs deleted in HEAD and modified in develop. Version develop of sys/valexpr/src/CstrDef.hs left in tree.
CONFLICT (modify/delete): sys/valexpr/src/ConstDefs.hs deleted in develop and modified in HEAD. Version HEAD of sys/valexpr/src/ConstDefs.hs left in tree.
Auto-merging sys/sort/src/Sort/Internal.hs
CONFLICT (content): Merge conflict in sys/sort/src/Sort/Internal.hs
Auto-merging sys/sort/src/Id.hs
Auto-merging sys/solve/test/TestVexprToSMT.hs
CONFLICT (content): Merge conflict in sys/solve/test/TestVexprToSMT.hs
Auto-merging sys/solve/test/TestSortDefToSMT.hs
Auto-merging sys/solve/test/TestSolvers.hs
CONFLICT (content): Merge conflict in sys/solve/test/TestSolvers.hs
Auto-merging sys/solve/test/TestSMTValue.hs
Auto-merging sys/solve/test/TestRecursiveFunction.hs
CONFLICT (content): Merge conflict in sys/solve/test/TestRecursiveFunction.hs
Auto-merging sys/solve/test/TestPushPop.hs
CONFLICT (content): Merge conflict in sys/solve/test/TestPushPop.hs
Auto-merging sys/solve/test/TestMultipleInstances.hs
CONFLICT (content): Merge conflict in sys/solve/test/TestMultipleInstances.hs
Auto-merging sys/solve/test/TestFuncDefToSMT.hs
CONFLICT (content): Merge conflict in sys/solve/test/TestFuncDefToSMT.hs
Auto-merging sys/solve/test/TestConstraint.hs
CONFLICT (content): Merge conflict in sys/solve/test/TestConstraint.hs
Auto-merging sys/solve/test/HelperVexprToSMT.hs
CONFLICT (content): Merge conflict in sys/solve/test/HelperVexprToSMT.hs
Auto-merging sys/solve/test/HelperFuncDefToSMT.hs
CONFLICT (content): Merge conflict in sys/solve/test/HelperFuncDefToSMT.hs
Auto-merging sys/solve/src/TXS2SMT.hs
CONFLICT (content): Merge conflict in sys/solve/src/TXS2SMT.hs
Auto-merging sys/solve/src/Solve.hs
CONFLICT (content): Merge conflict in sys/solve/src/Solve.hs
Auto-merging sys/solve/src/SMTInternal.hs
CONFLICT (content): Merge conflict in sys/solve/src/SMTInternal.hs
Auto-merging sys/solve/src/SMT2TXS.hs
CONFLICT (content): Merge conflict in sys/solve/src/SMT2TXS.hs
Auto-merging sys/solve/src/RandTrueBins.hs
CONFLICT (content): Merge conflict in sys/solve/src/RandTrueBins.hs
Auto-merging sys/solve/src/RandPartition.hs
CONFLICT (content): Merge conflict in sys/solve/src/RandPartition.hs
Auto-merging sys/solve/src/RandIncrementChoice.hs
CONFLICT (content): Merge conflict in sys/solve/src/RandIncrementChoice.hs
Auto-merging sys/solve/src/FreeVar.hs
Auto-merging sys/solve/solve.cabal
CONFLICT (content): Merge conflict in sys/solve/solve.cabal
Auto-merging sys/server/server.cabal
CONFLICT (content): Merge conflict in sys/server/server.cabal
Removing sys/front/test/Main.hs
CONFLICT (modify/delete): sys/front/src/TxsHappy.y deleted in develop and modified in HEAD. Version HEAD of sys/front/src/TxsHappy.y left in tree.
Removing sys/front/src/TxsAlex.x
CONFLICT (modify/delete): sys/front/front.cabal deleted in develop and modified in HEAD. Version HEAD of sys/front/front.cabal left in tree.
Auto-merging sys/defs/src/XmlFormat.hs
CONFLICT (content): Merge conflict in sys/defs/src/XmlFormat.hs
Auto-merging sys/defs/src/TxsUtils.hs
CONFLICT (content): Merge conflict in sys/defs/src/TxsUtils.hs
Auto-merging sys/defs/src/TxsShow.hs
CONFLICT (content): Merge conflict in sys/defs/src/TxsShow.hs
Auto-merging sys/defs/src/StdTDefs.hs
CONFLICT (content): Merge conflict in sys/defs/src/StdTDefs.hs
Removing sys/defs/src/Params.hs
Auto-merging sys/defs/src/Ident.hs
Auto-merging sys/defs/src/FuncTable.hs
CONFLICT (content): Merge conflict in sys/defs/src/FuncTable.hs
Removing sys/defs/src/BehExprDefs.hs
Auto-merging sys/defs/defs.cabal
CONFLICT (content): Merge conflict in sys/defs/defs.cabal
Auto-merging sys/coreenv/coreenv.cabal
Auto-merging sys/core/src/TxsCore.hs
Auto-merging sys/bexpr/src/VarEnv.hs
Auto-merging sys/bexpr/src/ProcId.hs
CONFLICT (content): Merge conflict in sys/bexpr/src/ProcId.hs
Auto-merging sys/bexpr/src/ProcDef.hs
Auto-merging sys/behaveenv/behaveenv.cabal
Auto-merging sys/behavedefs/src/BTree.hs
Auto-merging sys/behavedefs/behavedefs.cabal
Auto-merging sys/behave/test/TestHelperFuncContent.hs
Removing sys/behave/src/Subst.hs
Auto-merging sys/behave/src/Expand.hs
CONFLICT (content): Merge conflict in sys/behave/src/Expand.hs
Auto-merging sys/behave/behave.cabal
CONFLICT (content): Merge conflict in sys/behave/behave.cabal
Auto-merging stack_linux.yaml
Auto-merging stack.yaml
Removing snap/snapcraft.yaml
Removing snap/hooks/install
Removing snap/data/.torxakis.yaml
Removing snap/README.md
Auto-merging ci/updateAndPushGhPagesFromDevelop.sh
CONFLICT (content): Merge conflict in ci/updateAndPushGhPagesFromDevelop.sh
Auto-merging .gitignore
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by pjljvandelaar** [comment]
@keremispirli Why is a user defined ADTDef of type `ADTDEF (Either Sort Name)`?
Can't it be just `ADTDEF Name` (i.e. user refer by name)
and `ADTDEF Sort` for all internal references (i.e. we refer by checked names)?

How could a user get hold of our internal references?
Originally we thought we had to share but with using hashing that wasn't needed anymore.
So we don't distribute them ...

**Comment by keremispirli** [comment]
I think we have tried `ADTDEF Name` and `ADTDEF Sort` but at some point (I think it was close to the compiler) we got stuck; the only choice was `ADTDEF (Either Sort Name)`. I remember writing the details down somewhere but can't find it now.

On the other hand, since now we have a new compiler it may indeed be possible to use something simpler.
